### PR TITLE
SDIT-643 Allow retry of failed prison events

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

**Review by petergphillips — DISMISSED** [review]
LGTM

File: .circleci/config.yml [hunk 1]
```diff
@@ -13,13 +13,47 @@ parameters:
...
+          - TMPDIR=/private
+      - image: gvenzl/oracle-xe:18-slim
+        environment:
+          - ORACLE_PASSWORD=test
+
+    environment:
+      _JAVA_OPTIONS: -Xmx256m -XX:ParallelGCThreads=2 -XX:ConcGCThreads=2 -Djava.util.concurrent.ForkJoinPool.common.parallelism=2 -Dorg.gradle.daemon=false -Dkotlin.compiler.execution.strategy=in-process -Dorg.gradle.workers.max=1
+    working_directory: ~/app
+    resource_class: medium
+
```
Comment (steverendell) [inline .circleci/config.yml:37]: is there a way of using the hmpps/localstack and also pulling in an extra docker image?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,38 @@
...
+colima start --arch x86_64 --memory 4 --network-address
+```
+
+Next run 
+
+`docker-compose -f docker-compose-test.yml up`
+
+Now, integration tests will run with these existing docker containers .
+
+Testcontainers will not start new containers on Mac as the port technique to detect an already running instance does not work under colima.
```
Comment (steverendell) [inline README.md:38]: This looks a little confused. I cant use testContainers atm as the test for whether the port is in use does not work under Colima. But a more sophisticated test should work

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/integration/LocalStackContainer.kt [hunk 3]
```diff
@@ -39,10 +43,11 @@ object LocalStackContainer {
   }
 
   private fun localstackIsRunning(): Boolean =
-    try {
-      val serverSocket = ServerSocket(4566)
-      serverSocket.localPort == 0
-    } catch (e: IOException) {
-      true
-    }
```
Comment (steverendell) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/integration/LocalStackContainer.kt:47]: Disable testContainers for now if using Colima

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/integration/OracleToTopicIntTest.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    await untilCallTo { getNumberOfMessagesCurrentlyOnPrisonEventQueue() } matches { it == 1 }
+
+    val receiveMessageResult = prisonEventQueueSqsClient.receiveMessage(prisonEventQueueUrl)
+    with(receiveMessageResult.messages.first()) {
+      assertThat(body.contains("""\"nomisEventType\":\"OFF_RECEP_OASYS\"""")).isTrue()
+      assertThat(body.contains("""\"eventType\":\"OFFENDER_MOVEMENT-RECEPTION\"""")).isTrue()
+      assertThat(body.contains("""\"bookingId\":1234567""")).isTrue()
+      assertThat(body.contains("""\"movementSeq\":4""")).isTrue()
+    }
+  }
```
Comment (steverendell) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/integration/OracleToTopicIntTest.kt:55]: Just a happy path test so far

File: .circleci/config.yml [hunk 5]
```diff
@@ -13,13 +13,47 @@ parameters:
...
+              curl -4 --connect-timeout 30 http://localhost:1521
+              ERR=$?
+              if [[ $ERR == 0 || $ERR == 52 ]]; then
+                RTN=0
+                break
+              fi
+              sleep 10
+            done
+            exit $RTN
+
```
[the remark this answers is not in the document]
Comment (andymarke) [reply .circleci/config.yml:56]: I thought there was a helper Circle Orb to this, where you just wait for a port to be active

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/service/AQService.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    val exceptionQueue = session.getQueue(tableOwner, EXCEPTION_QUEUE_NAME)
+    queueConnection.start()
+    try {
+      val dpsQueue = session.getQueue(tableOwner, queueSimpleName)
+      val messageConsumer = session.createConsumer(
+        exceptionQueue,
+        "JMSMessageID in ${join(ids)}", // MSGID column
+      )
+      val messageProducer = session.createProducer(dpsQueue)
+      while (true) {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/service/AQService.kt:38]: could you move `val message = messageConsumer.receiveNoWait() `  into the `while` so the `while (true)` reads better?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/service/AQService.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  private val sqlRepository: SqlRepository,
+  private val queueConnectionFactory: QueueConnectionFactory,
+  @Value("\${source.queue.name}") val queueName: String,
+) {
+  fun requeueExceptions() {
+    val queueDetails = queueName.split(".")
+    val tableOwner = queueDetails[0]
+    val queueSimpleName = queueDetails[1]
+
+    val ids = sqlRepository.getExceptionMessageIds()
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerevents/service/AQService.kt:25]: I think SQS library has telemetry that logs the number of messages it retires, make take a look and make the logging identical to the SQS library?

**Review by andymarke — APPROVED** [review]
LGTM

File: .circleci/config.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (steverendell) [reply .circleci/config.yml:56]: There is but it expects curl to return a 0 status whereas for oracle annoyingly it returns 52

**Pull request merged**